metrics/registry.go

```diff
@@ -89,3 +99,32 @@ func (r *Registry) MustNewMetric(name string, typ MetricType, t ...ValueType) *M
 func (r *Registry) Get(name string) *Metric {
 	return r.metrics[name]
 }
+
+// BranchTagSetRoot creates a new TagSet starting from the root.
+func (r *Registry) BranchTagSetRoot() *TagSet {
+	return &TagSet{tags: r.tagSetRoot}
+}
+
+// BranchTagSetRootWith creates a new TagSet starting from the root and
+// adds all the key-value pairs provided with the map to the new TagSet.
+func (r *Registry) BranchTagSetRootWith(m map[string]string) *TagSet {
```

Comment:
I have a few of concerns about these names:
1. "TagSetRoot" is unfortunately very easy to misread as `Branch_Tag_SetRoot`, i.e. set the root of the branch tag (whatever that means), not as `Branch_TagSet_Root`... So I suggest replacing `TagSet` with just `Tags`, or changing the order from `TagSetRoot` to `RootTagSet`
2. Branch is most often a noun, not a verb, so "BranchTag..." is a bit confusing. The `BranchOut()` method in `TagSet` is much clearer :+1:

Finally, I am not sure the second method, `BranchTagSetRootWith()` (however it's named), should be in the `Registry`, it isn't in any way specific to the root, right? Why not have it just in the `metrics.TagSet`?

Here are my concrete suggestions. Rename the `BranchTagSetRoot()` method to `RootTagSet()` and move the second method to `TagSet` and rename it to `BranchOutWith()`
```suggestion
// RootTagSet creates and returns a new TagSet, starting from the common global root.
func (r *Registry) RootTagSet() *TagSet {
	return &TagSet{tags: r.tagsRoot}
}

// BranchOutWith returns a new TagSet by branching out from the receiver
// and adding the key-value pairs provided with the map to the new TagSet.
func (tg *TagSet) BranchOutWith(m map[string]string) *TagSet {
```

It will make some one-liners in the tests from this PR and https://github.com/grafana/k6/pull/2633 a bit longer, but I'd argue it will also make them more readable. For example, these lines:
```go
piState.Registry.BranchTagSetRootWith(options.RunTags)

piState.Registry.BranchTagSetRootWith(map[string]string{"a": "1"}).SampleTags(),
```

will become these:
```go
piState.Registry.RootTagSet().BranchOutWith(options.RunTags)

piState.Registry.RootTagSet().BranchOutWith(map[string]string{"a": "1"}).SampleTags(),
```

Reply:
hmm or maybe even `TagsRoot()`:
```go
piState.Registry.TagsRoot().BranchOutWith(options.RunTags)

piState.Registry.TagsRoot().BranchOutWith(map[string]string{"a": "1"}).SampleTags(),
```
:thinking:

Reply:
`TagsRoot` seems nicer

Reply:
>Here are my concrete suggestions. Rename the BranchTagSetRoot() method to RootTagSet()

It is fine to me to have a better name but I think the `BranchOutTagsRoot()` or in a different order is better, because it makes it visible that it is already branched out and you don't need to branch after. Because with `RootTagSet()` it seems legit to think I have to do `RootTagSet().BranchOut()`.

```go
piState.Registry.RootTagSet().BranchOutWith(options.RunTags)
```

Also, for the same reason, we are allocating `metrics.TagSet` twice for `RootTagSet()` and for `BranchOutWith(options.RunTags)`. We could resolve this by storing a centralised `metrics.TagSet` instead of the `atlas.Node` but we said we prefer the node for safety.

>Finally, I am not sure the second method, BranchTagSetRootWith() (however it's named), should be in the Registry, it isn't in any way specific to the root, right? Why not have it just in the metrics.TagSet?

>and move the second method to TagSet and rename it to BranchOutWith()

:-1: I think this will increase the risk we use TagSet in a bad form. I think it's root specific logic because the good case where we should use a `BranchOutWith` is from the root. For the rest, we should design the code in `branch := TagSet.BranchOut(); branch.AddTag(k,v)`. In this way, it forces us to think about to have a consistent order and reusing the order from the passed structure (e.g like we did for the js modules reusing the `goja.Object`).

Reply:
Ah, very good points :+1: :man_facepalming: I am not too worried about the performance implications of branching twice from the root, since that will only happen a few times in the actual k6 run, but you are right that it's a bad example. And, while you are right that the only good case for `BranchOutWith` is from the root, I kind of still don't think this method should be in the Registry :thinking: 

So, new suggestion :sweat_smile: 
```suggestion
// TagsRoot creates and returns a new TagSet, starting from the common global root.
func (r *Registry) TagsRoot() *TagSet {
	return &TagSet{tags: r.tagsRoot}
}

// SortAndAddTags sorts the given tags by their keys and adds them to the 
// current tag set one by one, without branching out. This is generally 
// discouraged and sequential usage of the AddTag() method should be 
// preferred and used whenever possible. 
// 
// The only place this method should be used is if we already have a 
// map[string]string of tags, e.g. with the test-wide --tags from the root,
// with scenario.tags, with custom per-reqeust tags from a user, etc.
// Then it's more efficient to sort their keys before we add them. If we don't,
// go map iteration happens in pseudo-random order and this will generate
// a lot of useless dead-end atlas Nodes.
func (tg *TagSet) SortAndAddTags(m map[string]string) *TagSet {
```

Reply:
We had a video call to resolve this and the notes are in https://github.com/grafana/k6/pull/2594#issuecomment-1205273413